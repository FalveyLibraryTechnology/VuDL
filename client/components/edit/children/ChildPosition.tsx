import React, { useState } from "react";
import TextField from "@mui/material/TextField";
import { useEditorContext } from "../../../context/EditorContext";
import { useFetchContext } from "../../../context/FetchContext";
import { getPositionInParentUrl } from "../../../util/routes";

export interface ChildPositionProps {
    pid: string;
    parentPid: string;
}

export const ChildPosition = ({ pid, parentPid }: ChildPositionProps): React.ReactElement => {
    const {
        state: { objectDetailsStorage },
        action: { clearPidFromChildListStorage, removeFromObjectDetailsStorage },
    } = useEditorContext();
    const {
        action: { fetchText },
    } = useFetchContext();
    const [statusMessage, setStatusMessage] = useState<string>("");

    // Important: this component assumes it will be called in a context where pid and parent pid
    // details are already loaded into the context; it will not cause a load to occur by itself.
    const loaded = Object.prototype.hasOwnProperty.call(objectDetailsStorage, pid);
    const parentLoaded = Object.prototype.hasOwnProperty.call(objectDetailsStorage, parentPid);
    const details = loaded ? objectDetailsStorage[pid] : {};
    const parentDetails = parentLoaded ? objectDetailsStorage[parentPid] : {};
    const parentSort = parentDetails.sortOn ?? "title";

    function findSequence(sequences: Array<string>): number | boolean {
        // We only need a sequence number if we're inside a sorted parent...
        if (parentSort === "custom") {
            for (let x = 0; x < sequences.length; x++) {
                const parts = sequences[x].split("#");
                if (parts[0] === parentPid) {
                    return parseInt(parts[1]);
                }
            }
        }
        return false;
    }

    const sequence = findSequence(details.sequences ?? []);
    async function saveChange(event: React.ChangeEvent<HTMLInputElement>) {
        const newValue = parseInt(event.target.value);
        if (newValue === sequence) {
            // Don't save if nothing has changed:
            return;
        }
        setStatusMessage("Saving...");
        const target = getPositionInParentUrl(pid, parentPid);
        const result = await fetchText(target, { method: "PUT", body: newValue });
        if (result === "ok") {
            // Clear and reload the cached object, since it has now changed!
            removeFromObjectDetailsStorage(pid);
            // Clear any cached lists belonging to the parent PID, because the
            // order has potentially changed!
            clearPidFromChildListStorage(parentPid);
        }
        setStatusMessage("");
    }
    if (statusMessage.length > 0) {
        return <>{statusMessage}</>;
    }
    return sequence === false ? (
        <></>
    ) : (
        <TextField
            sx={{ width: "5em", marginRight: 1 }}
            size="small"
            aria-label={`Position in ${parentPid}`}
            defaultValue={sequence}
            onBlur={saveChange}
        />
    );
};

export default ChildPosition;
