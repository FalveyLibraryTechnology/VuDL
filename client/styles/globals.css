:root,
[color-scheme="light"] {
    --body-bg: #fafbfc;
    --heading-color: #4d4d4d;
    --link-color: #008aed;

    --btn-bg: #485a6c;
    --btn-text: #ffffff;
    --btn-border: #275575;
    --btn-focus-bg: #3498db;

    --btn-primary-bg: #27ae60;
    --btn-primary-text: #ffffff;
    --btn-primary-border: #275575;
    --btn-primary-focus-bg: #155724;

    --btn-danger-bg: red;
    --btn-danger-text: #ffffff;
    --btn-danger-border: #a10705;
    --btn-danger-focus-bg: #a10705;

    --btn-logout-bg: #1abc9c;
    --btn-logout-text: #ffffff;
    --btn-logout-border: #16a085;
    --btn-logout-focus-bg: #1abc9c;

    --btn-toggle-bg: #7f8c8d;
    --btn-toggle-text: #ffffff;
    --btn-toggle-border: #34495e;
    --btn-toggle-focus-bg: #53626f;

    --group-border: #53626f;

    --input-border: #6a737d;
    --input-focus-border: #2188ff;

    --label-hover: #c8e1ff;

    --page-list-bg: #ffffff;
    --page-list-border: #c1c1c1;

    --thumbnail-bg: #e1e1e1;
    --thumbnail-focus-bg: #51aded;
    --thumbnail-number-bg: #ffffff;
    --thumbnail-number-text: #000000;
    --thumbnail-preview-border: #000000;

    --tree-item-focus-bg: #000000;
    --tree-item-focus-text: #ffffff;

    --shadow-gray: #c8e1ff;
}

[color-scheme="dark"] {
}
