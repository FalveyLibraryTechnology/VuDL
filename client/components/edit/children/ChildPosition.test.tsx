import React from "react";
import { beforeEach, describe, expect, it, jest } from "@jest/globals";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import ChildPosition from "./ChildPosition";

const mockUseEditorContext = jest.fn();
jest.mock("../../../context/EditorContext", () => ({
    useEditorContext: () => {
        return mockUseEditorContext();
    },
}));

const mockUseFetchContext = jest.fn();
jest.mock("../../../context/FetchContext", () => ({
    useFetchContext: () => {
        return mockUseFetchContext();
    },
}));

describe("ChildPosition", () => {
    let editorValues;
    let fetchContextValues;
    const pid = "foo:123";
    const parentPid = "foo:122";
    beforeEach(() => {
        editorValues = {
            state: {
                objectDetailsStorage: {},
            },
            action: {
            },
        };
        mockUseEditorContext.mockReturnValue(editorValues);
        fetchContextValues = {
            action: {
                fetchText: jest.fn(),
            },
        };
        mockUseFetchContext.mockReturnValue(fetchContextValues);
    });

    it("renders correctly when no sort or sequence exists", () => {
        editorValues.state.objectDetailsStorage[pid] = { pid };
        editorValues.state.objectDetailsStorage[parentPid] = { pid: parentPid };
        const wrapper = shallow(<ChildPosition pid={pid} parentPid={parentPid} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("renders correctly when a sequence exists and parent is not custom sorted", () => {
        const sequences = [`${parentPid}#73`];
        editorValues.state.objectDetailsStorage[pid] = { pid, sequences };
        editorValues.state.objectDetailsStorage[parentPid] = { pid: parentPid, sortOn: "title" };
        const wrapper = shallow(<ChildPosition pid={pid} parentPid={parentPid} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("renders correctly when a sequence exists and parent is custom sorted", () => {
        const sequences = [`${parentPid}#73`];
        editorValues.state.objectDetailsStorage[pid] = { pid, sequences };
        editorValues.state.objectDetailsStorage[parentPid] = { pid: parentPid, sortOn: "custom" };
        const wrapper = shallow(<ChildPosition pid={pid} parentPid={parentPid} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
