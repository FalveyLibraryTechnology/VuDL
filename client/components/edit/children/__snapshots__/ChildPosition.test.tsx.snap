// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChildPosition renders correctly when a sequence exists and parent is custom sorted 1`] = `
<div
  aria-label="Position in foo:122"
  className="MuiFormControl-root MuiTextField-root css-laf9zm-MuiFormControl-root-MuiTextField-root"
>
  <div
    className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
    onClick={[Function]}
  >
    <input
      aria-invalid={false}
      autoFocus={false}
      className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
      defaultValue={73}
      disabled={false}
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      required={false}
      type="text"
    />
    <fieldset
      aria-hidden={true}
      className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
    >
      <legend
        className="css-ihdtdm"
      >
        <span
          className="notranslate"
        >
          â€‹
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;
